qa/tasks/rsync.py

```diff
@@ -0,0 +1,276 @@
+                    self.mount_point = tmp
+                    self.my_mnt = j
+                if (not self.mount_point) and (not flag):
+                    self.mount_point = tmp
+                    self.my_mnt = j
+                    flag = True
+        else:
+            assert len(self.ctx.mounts.items()) > 0, 'No mount available asserting rsync'
+
+        #Set source directory for rsync
+        if self.config.get('data_dir'):
+            self.work_unit = True
+            self.source_dir = misc.get_testdir(self.ctx) + '/mnt.{}'.format(self.mount_point) + \
+                              '/{}/'.format(self.config.get('data_dir'))
+        else:
+            self.data_dir = misc.get_testdir(self.ctx) + '/mnt.{}/'.format(self.mount_point) + 'source'
+            self.source_dir = self.data_dir + '/subdir'
+
+    def _run(self):
+        try:
+            self.do_rsync()
+        except:
+            # Log exceptions here so we get the full backtrace (it's lost
+            # by the time someone does a .get() on this greenlet)
+            self.logger.exception("Exception in do_rsync:")
+            raise
+
+    def stop(self):
+        self.stopping.set()
+
+    #Function to check directory exists.
+    def check_if_dir_exists(self, path):
+        """
+            Call this to check stat of directory.
+            :return: True if directory exists.
+                     False if not.
+        """
+        try:
+            self.my_mnt.stat(path)
+            return True
+        except Exception, e :
+            logging.error(e)
+            return False
+
+    def do_rsync(self):
+
+        iteration = 0
+        finished = False
+
+        # Create destination directory
+        self.my_mnt.run_shell(["mkdir", "rsyncdir"])
+
+        if not self.work_unit:
+            # Create a data directory, sub directory and rsync directory
+            self.my_mnt.run_shell(["mkdir", "{}".format(self.data_dir)])
+            self.my_mnt.run_shell(["mkdir", "{}".format(self.source_dir)])
+            should_stop = False
+
+        #Check for source directory exists
+        while not self.check_if_dir_exists(self.source_dir):
+            time.sleep(5) # if source dorectory not exists wait for 5s and poll
+            iteration += 1
+            if iteration > 5:
+                assert self.check_if_dir_exists(self.source_dir), 'assert, source Directory does not exists'
+
+        # Start observing the event started by workunit task.
+        if self.work_unit:
+            should_stop = self.ctx.workunit_state.start_observing()
+
+        while not (should_stop or self.stopping.is_set()):
+
+            # rsync data from snapshot. snap is created using workunit IO data
+            if self.work_unit and self.snap_enable:
+
+                snap_shot = self.source_dir + '.snap/snap' + '{}'.format(iteration)
+
+                # Create Snapshot
+                self.my_mnt.run_shell(["mkdir", "{}".format(snap_shot)])
```

Comment:
Let's refactor this into a Filesystem method that takes a path and snapshot name.

Reply:
Ack. Done!

Reply:
@batrick It should be a method ( create_snapshot() ) in mount.py right? Because mount object is necessary for running mkdir.

Reply:
Yes, that's right.